mautic/mautic
# SMS enhancements - support callback with received & read & failed notifications

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
This PR allow any text messages transport add delivered, read or failed notification.
Extend sms_messages table with: deliveredCount, readCount, failedCount column
Extend sms_messages_stat table with is_delivered, is_read, is_failed  column
All this columns are optional and disabled by default. Every transport can enabled it.

This PR is follower of new 2.16 feature Add support for sms replies as a campaign decision https://github.com/mautic/mautic/pull/7449

For DEVS:

Every SMS transport can implements TransportSettingsInterface and implement method enabledSettings with  array of any constant

TransportSettingsInterface::STAT_DELIVERED
TransportSettingsInterface::STAT_READED
TransportSettingsInterface::STAT_FAILED

This add new stats to column and graphs in details

![image](https://user-images.githubusercontent.com/462477/77453503-a2914300-6df7-11ea-8fc5-6503c32f79f0.png)

![image](https://user-images.githubusercontent.com/462477/77453553-b2108c00-6df7-11ea-88e7-4c59858372b9.png)

Usage:

Example Transport

https://gist.github.com/kuzmany/995313c52f82bbb245152404f0c0d53d#file-exampletransport-php-L14

Example of Transport Callback

https://gist.github.com/kuzmany/1635e2684b4728cabb26e0689b3dbf5c

Dispatch Reply or Delivery events

https://gist.github.com/kuzmany/01001d7e49db8e9ddc00bbaa0592e314

#### How to test it:
You have to register a webhook with the service provider. For Twilio, it's done under Messaging for the phone number. The callback is `https://your.mautic.net/sms/twilio/callback`

1. Configure a SMS gateway 
2. Create a campaign with a "send text message" action then a "sends a text message" decision and do something as a result
3. Reply to a SMS 
4. The action attached to the decision should have been executed



#### List backwards compatibility breaks:
1. Refactor app\bundles\SmsBundle\Helper\ReplyHelper.php to app\bundles\SmsBundle\Helper\CallbackHelper.php
2. Remove getMessage() from app\bundles\SmsBundle\Callback\CallbackInterface.php
3. Add getEvent() to app\bundles\SmsBundle\Callback\CallbackInterface.php - should return ReplyEvent() or DeliveryEvent() object

If 3th party addon use that, need change it
